testhelper/tests.go

```diff
@@ -293,23 +300,30 @@ func (options *TestOptions) RunTestUpgrade() (*terraform.PlanStruct, error) {
 		_, resultErr = terraform.InitAndApplyE(options.Testing, options.TerraformOptions)
 		assert.Nilf(options.Testing, resultErr, "Terraform Apply on MASTER branch has failed")
 
+		// Get the path to the state file
+		statePath := path.Join(options.TerraformDir, "terraform.tfstate")
+
+		// Copy the state file to the temporary directory
+		errCopyState := common.CopyFile(statePath, path.Join(tempDir, "terraform.tfstate"))
+
 		// Only proceed to upgrade Test of master branch apply passed
-		if resultErr == nil {
+		if resultErr == nil && assert.Nil(options.Testing, errCopyState, fmt.Sprintf("Error copying state: %s", errCopyState)) {
+
 			logger.Log(options.Testing, "Attempting Git Checkout PR Branch:", prRef.Name(), "-", prRef.Hash())
 			// checkout the HASH of original (PR) branch.
 			// NOTE: in automation the original checkout branch is detached and points to the pseudo merge of the PR.
 			//       These detached merge branches report their name as "HEAD" which is not a suitable checkout value.
 			//       The solution here is to do this final checkout on the HASH of the original branch which will work
 			//       with both detached and normal branches.
 			resultErr = w.Checkout(&git.CheckoutOptions{
-				Hash: prRef.Hash(),
-				Keep: true}) // Keep true so we do not blow away the state file
+				Hash:  prRef.Hash(),
+				Force: true})
 			assert.Nilf(options.Testing, resultErr, "Could Not Checkout PR Branch")
 
-			_, stateErr := os.Stat(options.TerraformDir + "/terraform.tfstate")
-			assert.False(options.Testing, os.IsNotExist(stateErr), "State file not found cannot compare plan for test")
+			// Copy the state file back to the original location
+			errCopyState := common.CopyFile(path.Join(tempDir, "terraform.tfstate"), statePath)
```

Comment:
For future proofing (like if they change go-git behavior again) will want to make sure that we would either ignore an error if the statefile still exists in directory, or make sure that we force the copy even if it exists.